[FLINK-19176] Add pluggable statefun payload serializer

galenwarren opened the pull request:
This PR implements support for pluggable message payload serializers. It adds a new MessageFactoryType -- WITH_CUSTOM_PAYLOADS -- and a new class (MessageFactoryKey) that combines together MessageFactoryType and the name of a class to be used for payload serialization. This class must implement MessagePayloadSerializer.

This is pretty straightforward change, mostly replacing uses of MessageFactoryType with MessageFactoryKey. The changes to the MessageTypeSerializer.Snapshot class are a bit interesting, in that the serialized format for the snapshot had to change, to include the name of the custom serializer class. The snapshot version number was bumped from 1 to 2, and the new version of the class can read snapshots of both version 1 and 2, which is confirmed via new unit tests in MessageTypeSerializerSnapshotTest.

Any feedback or suggestions are welcome, and thanks to @igalshilman for pointing me in the right direction here.

tzulitai commented:
@galenwarren before we start reviewing this, @galenwarren can you squash your commits and rebase on latest master?

galenwarren commented:
@tzulitai 

Yes, of course. Sorry about that. Is this better?

galenwarren commented:
Hi, I just wanted to check if you needed anything from me at this point? 

I see a conflict has emerged on ```statefun-flink/statefun-flink-core/src/test/java/org/apache/flink/statefun/flink/core/StatefulFunctionsConfigTest.java```, let me know if you need me to do something there. Thanks.

igalshilman reviewed (changes requested):
Hi,
I’ve made the first pass on the PR and it looks really great, including the serializer migration path!
So really great job there!

I have few comments for you to consider, and I will make another pass tomorrow.

Regarding the merge conflict, can you please rebase on master one more time please? (There are no expect changes to the configuration class right now)

Thanks!

galenwarren commented on statefun-flink/statefun-flink-core/src/main/java/org/apache/flink/statefun/flink/core/message/MessageTypeSerializer.java:
```diff
@@ -101,45 +102,67 @@ public int hashCode() {
+    }
+
+    // used in unit tests
+    MessageFactoryKey getMessageFactoryKey() {
```
Done

galenwarren commented on statefun-flink/statefun-flink-core/src/test/java/org/apache/flink/statefun/flink/core/message/JavaPayloadSerializer.java:
```diff
@@ -0,0 +1,50 @@
+package org.apache.flink.statefun.flink.core.message;
```
Fixed

galenwarren commented on statefun-flink/statefun-flink-core/src/main/java/org/apache/flink/statefun/flink/core/StatefulFunctionsConfigValidator.java:
```diff
@@ -61,4 +64,26 @@ private static void validateParentFirstClassloaderPatterns(Configuration configu
+        configuration.get(StatefulFunctionsConfig.USER_MESSAGE_CUSTOM_PAYLOAD_SERIALIZER_CLASS);
+
+    if (factoryType == MessageFactoryType.WITH_CUSTOM_PAYLOADS) {
+      if (StringUtils.isNullOrWhitespaceOnly(customPayloadSerializerClassName)) {
```
A question here -- as it stands now, the USER_MESSAGE_CUSTOM_PAYLOAD_SERIALIZER_CLASS config option has no default value, so if not set, its value will be null. So I believe I'd have to first check for null and then check length on the trimmed string. I'm happy to do that, but I thought I'd check first ...

igalshilman replied:
I think that you are right, let's leave that as it is 👍

galenwarren replied:
OK will do

galenwarren commented on statefun-flink/statefun-flink-core/src/main/java/org/apache/flink/statefun/flink/core/message/MessageFactoryKey.java:
```diff
@@ -0,0 +1,43 @@
+    return this.type;
+  }
+
+  public String getCustomPayloadSerializerClassName() {
```
As it stands now, this method gets called from MessageTypeSerializer.Snapshot.writeSnapshot, for all types of message serializers, not just custom ones. For non-custom ones, a null is returned and this is the value that is written into the snapshot, after the MesssageFactoryType. So, the version 2 serialized format is always a MessageFactoryType value and then the custom payload serializer class name (string, which can be null), for all types of serializers.

The writeSnapshot method could be changed to only call getCustomPayloadSerializerClassName in the event that the serializer is custom, and then to either 1) force a null serializer class name to be written to the snapshot or 2) skip the write of the serializer class name altogether in the non-custom case. This would make the read/writeSnapshot code and testing a bit more complicated but would definitely be doable, would one of those options be better?

igalshilman replied:
That is a good point!
In that case, what do you think about making this method return an `Optional<String>`,
and version 2 would write an extra byte to indicate if the optional is present or not.

Something along the lines:
```
if (className.isPresent()) {
   out.writeBooolean(true);
   out.writeUtf8String(className.get());
} else {
   out.writeBooolean(false);
}
```

galenwarren replied:
For some reason, before I couldn't comment here, but now I can. Weird glitch. Anyway, I left a top-level comment/question for you on this one, just FYI.

igalshilman replied:
> 
> So the serialized format would already seem to be what you've requested! But if you'd prefer it get written without using `StringUtils.writeNullableString`, I'm happy to make that change.

Oh I see. My original thinking here was to prevent a future NPE by someone who would expect (mistakenly) that `messageFactoryKey.getCustomPayloadSerializerClassName()`  would always return a class name. Making the method return `Optional` would signal to the user of that class that it is not necessarily true.

And regarding `StringUtils` I'd rather to keep that logic local to the `Snapshot` class, to protect from accidental change in `StringUtils`.

galenwarren replied:
Great -- understood, I'll make that change and push a new version.

galenwarren replied:
These changes are in the code I just pushed, please let me know if it looks OK or if you need anything else. Thanks.

galenwarren commented:
@igalshilman 

Thanks for the review! I've rebased on master and fixed a couple of the issues you identified, and I have questions on the other two, in the comments.

igalshilman commented on statefun-flink/statefun-flink-core/src/main/java/org/apache/flink/statefun/flink/core/message/MessageFactoryKey.java:
```diff
@@ -0,0 +1,43 @@
+import java.io.Serializable;
+import java.util.Objects;
+
+public class MessageFactoryKey implements Serializable {
```
Can this be final?

galenwarren replied:
Yes, and it will be included in the next push

igalshilman commented on statefun-flink/statefun-flink-core/src/main/java/org/apache/flink/statefun/flink/core/message/MessageFactoryKey.java:
```diff
@@ -0,0 +1,43 @@
+  private final String customPayloadSerializerClassName;
+
+  private MessageFactoryKey(MessageFactoryType type, String customPayloadSerializerClassName) {
+    this.type = type;
```
Can this be: `this.type = Objects.requireNonNull(type);`

galenwarren replied:
Yes, will be fixed in next push

igalshilman reviewed (commented):
Hi,
I've went over the changes again, and I think that they are almost good to go!
I've left few minor comments but other than that I think that we are good to merge.

Thanks again for working on this!

galenwarren commented:
@igalshilman Thanks!

I'll take a look at the others shortly, but wanted to quickly reply regarding the question about Optional<String> and the serialized snapshot format. (I can't reply to the comment itself for some reason ...) What you propose is absolutely doable. I just wanted to make sure you're aware that the suggested code is essentially identical to the implementation of StringUtils.writeNullableString, which is what is currently used to write the version 2 snapshot in MessageTypeSerializer.Snapshot.writeSnapshot, i.e.:

```
    @Override
    public void writeSnapshot(DataOutputView dataOutputView) throws IOException {

      // version 1
      dataOutputView.writeUTF(messageFactoryKey.getType().name());

      // added in version 2
      StringUtils.writeNullableString(
          messageFactoryKey.getCustomPayloadSerializerClassName(), dataOutputView);
    }
```
... and ...

```
    public static void writeNullableString(@Nullable String str, DataOutputView out) throws IOException {
        if (str != null) {
            out.writeBoolean(true);
            writeString(str, out);
        } else {
            out.writeBoolean(false);
        }
    }
```

So the serialized format would already seem to be what you've requested! But if you'd prefer it get written without using ```StringUtils.writeNullableString```, I'm happy to make that change.

tzulitai reviewed (approved):
Thanks a lot for the contribution @galenwarren! This is really great and thought-out work.
LGTM, +1 to merge after @igalshilman's last few comments are addressed.

galenwarren commented:
@tzulitai  Thanks! I'm excited to be able to make a contribution here.

I pushed an update yesterday that addresses all of @igalshilman 's remaining comments except the one about writeSnapshot and the Snapshot serialization format, which I have a question about (above). Depending on the answer to that, I might have a bit more work to do, or we might be done.

Also, a question about process -- am I supposed to "Resolve Conversation" on items I've taken care of, or am I supposed to leave those open for someone else to resolve? Just not sure what the protocol is.

Thanks both of you for your help on this.

tzulitai commented:
> Also, a question about process -- am I supposed to "Resolve Conversation" on items I've taken care of, or am I supposed to leave those open for someone else to resolve? Just not sure what the protocol is.

You can resolve them yourself if you find them to be in a resolved state already.
Either way, committers will usually manually check that there are no outstanding discussions before merging.

galenwarren commented:
Great -- I've resolved the ones that are done and will work on the remaining item today. Thanks.

galenwarren commented:
The remaining item is resolved with the latest push.

igalshilman reviewed (approved):
Thanks a lot for addressing my comments. The PR looks really good, great job!

galenwarren commented:
@igalshilman 

Great! Thanks for your help and the kind words. So does that mean we're done?

Also, I was wondering if you know when this might make it into a release. No hurry, I can use my fork until then, just wondering.

tzulitai commented:
LGTM, merging this now to `master`!

> Also, I was wondering if you know when this might make it into a release. No hurry, I can use my fork until then, just wondering.

Since the feature freeze for 2.2 already happened, and this counts as a new feature, this will have to be included only in the next release (2.3).
The release cycles for StateFun are relatively shorter than Flink, so you can expect this to be officially available within 1~2 months when the next release is out.

tzulitai commented:
@galenwarren one comment about addressing PR reviews, for consideration in future contributions:
I find it best (and also recommended by the Flink contribution guide) to address PR comments with follow-up commits on the PR branch, instead of squashing + rebasing + force-pushing.
That way changes since the last review will be separate commits and will allow the reviewer to more easily track the changes.

Usually committers will be rebasing + squashing the follow-up commits when merging for you.

galenwarren commented:
>I find it best (and also recommended by the Flink contribution guide) to address PR comments with follow-up commits on the PR branch, instead of squashing + rebasing + force-pushing.

Thanks for the feedback -- will do that next time.
